File: typescript/deque_reference.ts
export class Deque<T> {
    private items: T[] = [];

    addFirst(item: T) { this.items.unshift(item); }
    addLast(item: T) { this.items.push(item); }

    removeFirst(): T | undefined { return this.items.shift(); }
    removeLast(): T | undefined { return this.items.pop(); }

    peekFirst(): T | undefined { return this.items[0]; }
    peekLast(): T | undefined { return this.items[this.items.length - 1]; }

    isEmpty(): boolean { return this.items.length === 0; }
}

function main() {
    let dq = new Deque<number>();
    dq.addFirst(1);
    dq.addLast(2);
    console.log(dq.removeFirst());
    console.log(dq.removeLast());
}

main();

File: javanotes/DequeReference.java
package javanotes;

import java.util.ArrayDeque;
import java.util.Deque;

public class DequeReference {
    public static void main(String[] args) {
        Deque<Integer> dq = new ArrayDeque<>(); // new LinkedList<>();

        // Add, Offer
        dq.addFirst(1);
        dq.addLast(2);
        dq.offerFirst(3); // offers are safer for capacity-restricted deques
        dq.offerLast(4);
        dq.addLast(5);
        System.out.println(dq);

        // Remove, Poll
        dq.removeFirst();
        dq.removeLast();
        dq.pollFirst(); // polls are safer if the deque might be empty
        dq.pollLast();
        System.out.println(dq);

        System.out.println(dq.getFirst());
        System.out.println(dq.peekLast()); // returns null if empty
    }
}
